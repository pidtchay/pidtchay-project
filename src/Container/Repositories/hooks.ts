/* eslint-disable jsdoc/require-returns */
import { useDispatch } from 'react-redux';
import { getRepositoriesData } from 'Store/repositories/api';
import { listRepositoriesSlice } from 'Store/repositories/slice';

/**
 * UseGithubRepos.
 */
export function useGithubRepos() {
    const dispatch = useDispatch();

    const { clean, fetched, fetching, fetchingError } = listRepositoriesSlice.actions;

    const fetchData = async (orgName: string) => {
        dispatch(fetching());
        return getRepositoriesData(orgName)
            .then((githubRepos) => {
                console.debug({ githubRepos });
                dispatch(fetched({ repositories: githubRepos }));
            })
            .catch((error) => {
                error.clientMessage = "Can't get repositories data";
                dispatch(fetchingError({ error }));
            });
    };

    const cleanData = () => {
        dispatch(clean());
    };

    return { fetchData, cleanData };
}
