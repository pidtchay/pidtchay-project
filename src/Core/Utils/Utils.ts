import { Loadable } from 'recoil';

export const isLoading = <T>(data: Loadable<T>): boolean => data.state === 'loading';

export const isSuccess = <T>(data: Loadable<T>): boolean => data.state === 'hasValue';

export const isError = <T>(data: Loadable<T>): boolean => data.state === 'hasError';
