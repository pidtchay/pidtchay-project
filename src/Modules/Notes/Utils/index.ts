const TITLE_MAX_LENGTH = 120;

export const getTitle = (value: string): string => {
    const lines = value.split('\n');
    return lines[0].length <= TITLE_MAX_LENGTH ? lines[0] : lines[0].substring(0, TITLE_MAX_LENGTH);
};

export const parseMarkdown = (markdownText: string): string => {
    const htmlText = markdownText
        .replace(/^### (.*$)/gim, '<h3>$1</h3>')
        .replace(/^## (.*$)/gim, '<h2>$1</h2>')
        .replace(/^# (.*$)/gim, '<h1>$1</h1>')
        // eslint-disable-next-line no-useless-escape
        .replace(/^\> (.*$)/gim, '<blockquote>$1</blockquote>')
        .replace(/```\n(.*)\n```/gim, '<code>$1</code>')
        .replace(/^\*\*(.*)\*\*/gim, '<b>$1</b>')
        .replace(/^\*(.*)\*/gim, '<i>$1</i>')
        .replace(/^~~(.*)~~/gim, '<s>$1</s>')
        .replace(/!\[(.*?)\]\((.*?)\)/gim, "<img alt='$1' src='$2' />")
        .replace(/\[(.*?)\]\((.*?)\)/gim, "<a href='$2'>$1</a>")
        .replace(/\n$/gim, '<br />');

    return htmlText.trim();
};
