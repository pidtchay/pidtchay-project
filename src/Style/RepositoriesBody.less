.repositories_body {
  padding: 15px;

  .repositories_body_up_button {
    height: 40px;
    width: 40px;
    line-height: 40px;
    border-radius: 4px;
    background-color: #1088e9;
    color: #fff;
    text-align: center;

    & > span {
      font-size: 20px;
    }
  }
}
