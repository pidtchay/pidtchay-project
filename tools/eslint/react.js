module.exports = {
    'react/no-did-update-set-state': 0,
    'react-hooks/exhaustive-deps': 0,
    'react/state-in-constructor': [2, 'never'],
    'react/destructuring-assignment': 0,
    'react/no-danger': 2,
    'react/prefer-stateless-function': 1,
    'react/no-array-index-key': 1,
    'react/jsx-props-no-spreading': 0,
    'react/static-property-placement': ['warn', 'property assignment', { // Указывает, где должны быть размещены статические свойства компонента React
      'childContextTypes': 'static getter',
      'contextTypes': 'static public field',
      'contextType': 'static public field',
      'displayName': 'static public field',
    }],
    'react/require-default-props': 1,
    'react/sort-comp': [1, { order: ['static-methods', 'lifecycle', 'everything-else', 'render']}],

}