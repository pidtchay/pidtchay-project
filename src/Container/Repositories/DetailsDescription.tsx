import { Descriptions, Tag } from 'antd';
import * as React from 'react';
import { useLiteralValue } from 'Utils/hooks';

interface IDetailsDescriptionProps {
    name: string;
    description: string;
    language: string;
}

export const DetailsDescription: React.FC<IDetailsDescriptionProps> = ({
    name,
    description,
    language
}) => {
    const { getValue: getLiteralValue } = useLiteralValue();
    return (
        <>
            <Descriptions
                title={getLiteralValue(
                    'Pages.GithubRepositories.Details.description'
                )}
            >
                <Descriptions.Item
                    label={getLiteralValue(
                        'Pages.GithubRepositories.Details.name'
                    )}
                >
                    {name}
                </Descriptions.Item>
                <Descriptions.Item
                    label={getLiteralValue(
                        'Pages.GithubRepositories.Details.description'
                    )}
                >
                    {description}
                </Descriptions.Item>
                <Descriptions.Item
                    label={getLiteralValue(
                        'Pages.GithubRepositories.Details.language'
                    )}
                >
                    <Tag color="geekblue">{language}</Tag>
                </Descriptions.Item>
            </Descriptions>
        </>
    );
};
