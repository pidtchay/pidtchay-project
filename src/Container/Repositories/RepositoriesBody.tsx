/* eslint-disable jsdoc/require-returns */
/* eslint-disable jsdoc/require-param */
import { Input, BackTop, Tooltip } from 'antd';
import * as React from 'react';
import { RocketOutlined } from '@ant-design/icons';
import { TableComponent } from 'Component/Common/TableComponent';
import { IRepositoryData } from 'Model/RepositoryData';
import { IColumnField } from 'Model/Table';
import style from 'Style/RepositoriesBody.less';
import { convertStringArrayToString } from 'Utils/common';
import { useLiteralValue } from 'Utils/hooks';

const { Search } = Input;

/**
 * @property {ILiterals} [literals] Language settings in the form of key-value. The text depends on the selected language.
 * @property {IColumnField[]} [columns] Table columns headers that will be visible to the user.
 * @property {IRepositoryData[]} [data] List of repositories for the selected organization name.
 * @property {Function} [onFetchRepositories] The function queries the list of repositories for the selected organization name.
 */
interface IRepositoriesBodyProps {
    columns: IColumnField[];
    data: IRepositoryData[];
    onFetchRepositories: (valuse: string) => void;
}

/**
 * Repositories Body Component.
 */
export const RepositoriesBody: React.FC<IRepositoriesBodyProps> = ({
    columns,
    data,
    onFetchRepositories
}) => {
    const { getValue: getLiteralValue } = useLiteralValue();
    const placeholder = convertStringArrayToString(
        getLiteralValue('Pages.GithubRepositories.search_placeholder')
    );
    return (
        <>
            <div>
                <Search
                    placeholder={placeholder}
                    enterButton="Search"
                    size="large"
                    onSearch={onFetchRepositories}
                />
            </div>
            <TableComponent columns={columns} dataSource={data || []} />
            <Tooltip title={getLiteralValue('ACTIONS.up')}>
                <BackTop className={style.repositories_body}>
                    <div className={style.repositories_body_up_button}>
                        <RocketOutlined />
                    </div>
                </BackTop>
            </Tooltip>
        </>
    );
};
