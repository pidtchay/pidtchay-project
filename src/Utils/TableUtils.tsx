import * as React from 'react';
import { IColumnField } from 'Model/Table';
import { IRepositoryData } from 'Model/RepositoryData';
import { Avatar, Tag } from 'antd';

enum ERenderType {
    TEXT= 'TEXT',
    AVATAR= 'AVATAR',
    LINK= 'LINK',
    TAG= 'TAG'
}

interface ISchema {
  description: string;
  name: string;
  type: ERenderType;
}

export const convertRepositoryDataToColumns = (data: ISchema[]): IColumnField[] => {
  if (!data) return [];
  return data.map((item) => {
    const column: IColumnField = {
      dataIndex: item?.name,
      key: item?.name,
      title: item?.name
    };

    if (item.type === ERenderType.AVATAR) {
      column.render = () => <Avatar src={item.description} />;
    }
    if (item.type === ERenderType.TAG) {
      column.render = () => <Tag color='geekblue'>{item.description}</Tag>;
    }

    return column;
  });
};

export const convertRepositoryDataToSchema = (data: IRepositoryData): ISchema[] => {
  if (!data) return [];
  return Object.keys(data).map((item) => {
    const schema: ISchema = {
      name: item,
      type: ERenderType.TEXT,
      description: data[item]
    };
    if (item === 'avatarImg') {
      schema.type = ERenderType.AVATAR;
    }
    if (item === 'owner' || item === 'language') {
      schema.type = ERenderType.TAG;
    }
    return schema;
  });
};
