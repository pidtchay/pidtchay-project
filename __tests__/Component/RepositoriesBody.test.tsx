import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { RepositoriesBody } from 'Container/Repositories/RepositoriesBody';
import { convertRepositoryDataToColumns, convertRepositoryDataToSchema } from 'Utils/TableUtils';

describe('render success',() => {
    it('should match the snapshot', () => {
        const repositoryData = [{
            avatarImg: 'http://test.com/avatar.jpeg',
            description: '',
            forks: 1,
            homepage: 'http://test.com/home',
            id: 1,
            language: 'javascript',
            name: 'ReactJS Todo list',
            owner: 'Monster corp',
            type: '',
            watchers: 5
        },
        {
            avatarImg: 'http://test.com/avatar2.jpeg',
            description: '',
            forks: 14,
            homepage: 'http://test2.com/home',
            id: 2,
            language: 'javascript',
            name: 'ReactJS Todo list2',
            owner: 'Robot corp',
            type: '',
            watchers: 35
        }];
        const columns = convertRepositoryDataToColumns(convertRepositoryDataToSchema(repositoryData[0]));
        const handleFetchData = () => jest.fn();

        const component = renderer.create(<RepositoriesBody columns={columns} data={repositoryData} onFetchRepositories={handleFetchData} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
});
