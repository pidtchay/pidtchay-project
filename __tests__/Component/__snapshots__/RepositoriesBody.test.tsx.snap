// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render success should match the snapshot 1`] = `
Array [
  <div>
    <span
      className="ant-input-search ant-input-search-enter-button ant-input-search-large ant-input-group-wrapper ant-input-group-wrapper-lg"
    >
      <span
        className="ant-input-wrapper ant-input-group"
      >
        <input
          className="ant-input ant-input-lg"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="input search org. name"
          style={null}
          type="text"
          value=""
        />
        <span
          className="ant-input-group-addon"
        >
          <button
            className="ant-btn ant-input-search-button ant-btn-primary ant-btn-lg"
            onClick={[Function]}
            onMouseDown={[Function]}
            type="button"
          >
            <span>
              Search
            </span>
          </button>
        </span>
      </span>
    </span>
  </div>,
  <div
    className="ant-table-wrapper"
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-table"
        >
          <div
            className="ant-table-container"
          >
            <div
              className="ant-table-content"
              onScroll={[Function]}
              style={Object {}}
            >
              <table
                style={
                  Object {
                    "tableLayout": "auto",
                  }
                }
              >
                <colgroup />
                <thead
                  className="ant-table-thead"
                >
                  <tr>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      avatarImg
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      description
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      forks
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      homepage
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      id
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      language
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      name
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      owner
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      type
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      watchers
                    </th>
                  </tr>
                </thead>
                <tbody
                  className="ant-table-tbody"
                >
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    onClick={[Function]}
                    style={Object {}}
                  >
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <span
                        className="ant-avatar ant-avatar-circle ant-avatar-image"
                        style={Object {}}
                      >
                        <img
                          onError={[Function]}
                          src="http://test.com/avatar.jpeg"
                        />
                      </span>
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      1
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      http://test.com/home
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <button
                        className="ant-btn ant-btn-link"
                        onClick={[Function]}
                        type="button"
                      >
                        1
                      </button>
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <span
                        className="ant-tag ant-tag-geekblue"
                        style={
                          Object {
                            "backgroundColor": undefined,
                          }
                        }
                      >
                        javascript
                      </span>
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      ReactJS Todo list
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <span
                        className="ant-tag ant-tag-geekblue"
                        style={
                          Object {
                            "backgroundColor": undefined,
                          }
                        }
                      >
                        Monster corp
                      </span>
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      5
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    onClick={[Function]}
                    style={Object {}}
                  >
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <span
                        className="ant-avatar ant-avatar-circle ant-avatar-image"
                        style={Object {}}
                      >
                        <img
                          onError={[Function]}
                          src="http://test.com/avatar.jpeg"
                        />
                      </span>
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      14
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      http://test2.com/home
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <button
                        className="ant-btn ant-btn-link"
                        onClick={[Function]}
                        type="button"
                      >
                        1
                      </button>
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <span
                        className="ant-tag ant-tag-geekblue"
                        style={
                          Object {
                            "backgroundColor": undefined,
                          }
                        }
                      >
                        javascript
                      </span>
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      ReactJS Todo list2
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <span
                        className="ant-tag ant-tag-geekblue"
                        style={
                          Object {
                            "backgroundColor": undefined,
                          }
                        }
                      >
                        Monster corp
                      </span>
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      35
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <ul
          className="ant-pagination ant-table-pagination ant-table-pagination-right"
          style={Object {}}
          unselectable="unselectable"
        >
          <li
            aria-disabled={true}
            className="ant-pagination-prev ant-pagination-disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={null}
            title="Previous Page"
          >
            <a
              className="ant-pagination-item-link"
              disabled={true}
            >
              <span
                aria-label="left"
                className="anticon anticon-left"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                  />
                </svg>
              </span>
            </a>
          </li>
          <li
            className="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex="0"
            title={1}
          >
            <a>
              1
            </a>
          </li>
          <li
            aria-disabled={true}
            className="ant-pagination-next ant-pagination-disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={null}
            title="Next Page"
          >
            <a
              className="ant-pagination-item-link"
              disabled={true}
            >
              <span
                aria-label="right"
                className="anticon anticon-right"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                  />
                </svg>
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>,
]
`;
